cache-at-ptr/shadow_opsem.py

```diff
@@ -0,0 +1,267 @@
+    elif outState.o[outState.r[q0]] > 0:
+        outState.o[outState.r[q0]] -= 1
+    outState.r.pop(q0, None)
+    return outState
+
+
+def brreg2reg(p0: Register,
+              q0: Register,
+              p1: Register,
+              inState: State) -> State:
+    """
+    `p1, q0 = brreg2reg p0`
+
+    """
+    if (q0 in inState.r.keys() or
+        p1 in inState.r.keys() or
+            p0 not in inState.r.keys() or
+            not inState.canMutBorrow(inState.r[p0])):
+        return inState
+    outState = copy.deepcopy(inState)
+    outState.pc = outState.pc + 1
+    outState.r[q0] = outState.r[p0]
+    # note: P1 is added immediately instead of waiting for
+    # q0 to die since we don't want to track aliases
+    # for this.
+    # Right now I don't know if this is a hindrance
+    # to proving equivalence with another machine.
+    outState.r[p1] = outState.r[p0]
+    outState.r.pop(p0, None)  # p0 is no longer used
+    return outState
+
+
+def brmem2reg(q0: Register,
+              pp0: Register,
+              inState: State) -> State:
+    """
+    m1, q0 = brmem2reg pp0, m0
+    """
+    if (q0 in inState.r.keys() or
+            pp0 not in inState.r.keys() or
+            not inState.canMutBorrow(inState.ptrptrto(pp0))):
+        return inState
+    outState = copy.deepcopy(inState)
+    outState.pc = outState.pc + 1
+    outState.r[q0] = outState.m[inState.r[pp0]]
+    return outState
+
+
+def brmem2reg_ro(q0: Register,
+                 pp0: Register,
+                 inState: State) -> State:
+    """
+    m1, q0 = brmem2reg_ro pp0, m0
+    """
+    if (q0 in inState.r.keys() or
+            pp0 not in inState.r.keys() or
+            not inState.canBorrow(inState.ptrptrto(pp0))):
+        return inState
+    outState = copy.deepcopy(inState)
+    outState.pc = outState.pc + 1
+    outState.r[q0] = inState.ptrptrto(pp0)
+    # ro borrow decrements ownership state of address
+    outState.o[inState.r[q0]] -= 1
+    return outState
+
+
+def mvreg2mem(pp0: Register,
+              p0: Register,
+              inState: State) -> State:
+    """
+    m1 = mvreg2mem pp0, p0, m0
+    """
+    if (pp0 not in inState.r.keys() or
+        p0 not in inState.r.keys() or
+            not inState.canMove(inState.r[p0])):
+        return inState
+    outState = copy.deepcopy(inState)
+    outState.pc = outState.pc + 1
+    outState.m[outState.r[pp0]] = outState.r[p0]
+    # no need to update ownership since it is keyed by address
+    outState.r.pop(p0, None)
+    return outState
+
+
+def mvmem2reg(p1: Register,
+              pp0: Register,
+              inState: State) -> State:
+    """
+    m1, p1 = mvmem2reg pp0, m0
+    """
+    if (pp0 not in inState.r.keys() or
+        p1 in inState.r.keys() or
+            not inState.canMove(inState.ptrptrto(pp0))):
+        return inState
+    outState = copy.deepcopy(inState)
+    outState.pc = outState.pc + 1
+    outState.r[p1] = outState.m[inState.r[pp0]]
```

Comment:
this should be `outState.r[p1] = outState.getPtrPtrTo(pp0)`. I will fix this in next iteration.